File: frontend/src/types/player.ts
// ABOUTME: TypeScript types for Player data matching backend Pydantic schemas
// ABOUTME: Ensures type safety between frontend and backend

export interface Player {
  id: number
  name: string
  email?: string
  trueskill_mu: number
  trueskill_sigma: number
  games_played: number
  wins: number
  losses: number
  created_at: string
  updated_at?: string
  is_active: boolean
  win_percentage: number
}

export interface PlayerCreate {
  name: string
  email?: string
}

export interface PlayerUpdate {
  name?: string
  email?: string
  is_active?: boolean
}

export interface PlayerListResponse {
  players: Player[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface RatingHistoryEntry {
  id: number
  player_id: number
  game_id: number
  trueskill_mu_before: number
  trueskill_sigma_before: number
  trueskill_mu_after: number
  trueskill_sigma_after: number
  rating_system: string
  created_at: string
  mu_change: number
  sigma_change: number
  conservative_rating_before: number
  conservative_rating_after: number
  conservative_rating_change: number
}

export interface RatingHistoryListResponse {
  rating_history: RatingHistoryEntry[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface PlayerRatingProgression {
  player_id: number
  player_name: string
  ratings: RatingHistoryEntry[]
}

export interface MultiPlayerRatingProgression {
  players: PlayerRatingProgression[]
  total_games: number
}

File: backend/app/schemas/player.py
# ABOUTME: Pydantic schemas for player data validation and serialization
# ABOUTME: Defines request/response models for player API endpoints

from datetime import datetime

from pydantic import BaseModel, EmailStr, Field


class PlayerBase(BaseModel):
    """Base player schema with common fields"""

    name: str = Field(..., min_length=1, max_length=100, description="Player name")
    email: EmailStr | None = Field(None, description="Player email address")


class PlayerCreate(PlayerBase):
    """Schema for creating a new player"""

    pass


class PlayerUpdate(BaseModel):
    """Schema for updating an existing player"""

    name: str | None = Field(
        None, min_length=1, max_length=100, description="Player name"
    )
    email: EmailStr | None = Field(None, description="Player email address")
    is_active: bool | None = Field(None, description="Whether player is active")


class PlayerResponse(PlayerBase):
    """Schema for player response data"""

    id: int
    trueskill_mu: float
    trueskill_sigma: float
    games_played: int
    wins: int
    losses: int
    win_percentage: float
    is_active: bool
    created_at: datetime
    updated_at: datetime | None

    class Config:
        from_attributes = True


class PlayerListResponse(BaseModel):
    """Schema for paginated player list response"""

    players: list[PlayerResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
